File: src/structural/proxy.java
public interface Internet {
    void connectTo(String host);
}

public class RealInternet implements Internet {
    @Override
    public void connectTo(String host) {
        System.out.println("Opened connection to " + host);
    }
}

public class ProxyInternet implements Internet {
    private static final List<String> bannedSites;
    private final Internet internet = new RealInternet();

    static {
        bannedSites = new ArrayList<>();
        bannedSites.add("banned.com");
    }

    @Override
    public void connectTo(String host) {
        if (bannedSites.contains(host)) {
            System.out.println("Access Denied!");
            return;
        }
        internet.connectTo(host);
    }
}

public static void main(String[] args) {
    Internet internet = new ProxyInternet();
    internet.connectTo("google.com");
    internet.connectTo("banned.com");
}

public static void main(final String[] arguments) {
    VideoDownloader videoDownloader = new RealVideoDownloader();
    videoDownloader.getVideo("geekific");
    videoDownloader.getVideo("geekific");
    videoDownloader.getVideo("likeNsub");
    videoDownloader.getVideo("likeNsub");
    videoDownloader.getVideo("geekific");
}

public interface VideoDownloader {
    Video getVideo(String videoName);
}

public class RealVideoDownloader implements VideoDownloader {
    @Override
    public Video getVideo(String videoName) {
        System.out.println("Connecting to https://www.youtube.com/");
        System.out.println("Downloading Video");
        System.out.println("Retrieving Video Metadata");
        return new Video(videoName);
    }
}

public class ProxyVideoDownloader implements VideoDownloader {
    private final Map<String, Video> videoCache = new HashMap<>();
    private final VideoDownloader downloader = new RealVideoDownloader();

    @Override
    public Video getVideo(String videoName) {
        if (!videoCache.containsKey(videoName)) {
            videoCache.put(videoName, downloader.getVideo(videoName));
        }
        return videoCache.get(videoName);
    }
}

File: src/structural/proxy.ts
interface Internet {
  connectTo(host: string): void;
}

class RealInternet implements Internet {
  connectTo(host: string): void {
    console.log("Opened connection to " + host);
  }
}

class ProxyInternet implements Internet {
  private static readonly bannedSites: string[] = [];
  private readonly internet = new RealInternet();

  static {
    this.bannedSites.push("banned.com");
  }

  connectTo(host: string): void {
    if (
      ProxyInternet.bannedSites.find((site) => {
        return site === host;
      })
    ) {
      console.log("Access Denied!", host);
      return;
    }
    this.internet.connectTo(host);
  }
}

function connect(): void {
  console.log("connect():");
  const internet = new ProxyInternet();
  internet.connectTo("google.com");
  internet.connectTo("banned.com");
  console.log("---");
}

connect();

class Video {
  constructor(protected name: string) {}
}

interface VideoDownloader {
  getVideo(videoName: string): Video | undefined;
}

class RealVideoDownloader implements VideoDownloader {
  getVideo(videoName: string): Video {
    console.log("RealVideoDownloader:");
    console.log("Connecting to https://www.youtube.com/");
    console.log("Downloading Video");
    console.log("Retrieving Video Metadata");
    console.log("---");
    return new Video(videoName);
  }
}

function downloadVideos(): void {
  console.log("downloadVideos():");
  const videoDownloader = new RealVideoDownloader();
  videoDownloader.getVideo("geekific");
  videoDownloader.getVideo("geekific");
  videoDownloader.getVideo("likeNsub");
  videoDownloader.getVideo("likeNsub");
  videoDownloader.getVideo("geekific");
  console.log("---");
}

downloadVideos();

class ProxyVideoDownloader implements VideoDownloader {
  private readonly videoCache = new Map<string, Video>();
  private readonly downloader = new RealVideoDownloader();

  getVideo(videoName: string): Video | undefined {
    console.log("ProxyVideoDownloader:");
    if (!this.videoCache.has(videoName)) {
      this.videoCache.set(videoName, this.downloader.getVideo(videoName));
    } else {
      console.log("cached video");
      console.log("---");
    }
    return this.videoCache.get(videoName);
  }
}

function downloadVideosProxy(): void {
  console.log("downloadVideosProxy():");
  const videoDownloader = new ProxyVideoDownloader();
  videoDownloader.getVideo("geekific");
  videoDownloader.getVideo("geekific");
  videoDownloader.getVideo("likeNsub");
  videoDownloader.getVideo("likeNsub");
  videoDownloader.getVideo("geekific");
  console.log("---");
}

downloadVideosProxy();
